contents/blog/null-handling-hogql.md

```diff
@@ -0,0 +1,82 @@
+
+## The problem with null in HogQL
+
+The issue with null in HogQL was that SQL is strict about nulls. If any value in an expression is null, then the entire expression is null. Here are two HogQL expressions showcasing why this is a problem:
+
+1. `sumIf(properties.money, event = 'no problems') + sumIf(properties.money - 10, event = 'some problems')` returns null if either of the sides of the equation are null. This  happens if there are no matching events (named `no problems` or `some problems`) or the property (`money`) is null.
+2. `concat('money: ', properties.money)` returns null if there's no `money` property, though users might expect either `'money: '`  or `'money: null'`
+
+PostHog’s existing solutions for handling null did not work with the full access HogQL provides. Specifically:
+
+- ClickHouse skips null during aggregation, but users could want to concat or check null values with HogQL.
+- PostHog previously treated nulls as strings, but this could include null values in counts when they shouldn’t be.
+
+At this stage in HogQL’s development, if a value could be null, users must handle it themselves. For example, users must wrap each `sumIf` function in the first example above in a `nullIf(x, 0)` like this:
+
+```
+nullIf(
+	sumIf(properties.money, event = 'no problems'), 
+	0
+) + nullIf(
+	sumIf(properties.money - 10, event = 'some problems'),
+	0
+)
+```
+
+This is not a good user experience as it is non-obvious they must do this and leads to a lot of extra code and checks.
+
+## Brainstorming potential solutions
+
+To solve the issue with nulls, Marius and Michael came up with a bunch of potential solutions, (none of which were inevitably chosen):
+
+1. **Returning a default value depending on type.** For example, returning `""` for strings, `0` for numbers, or `false` for booleans. This didn’t handle the expression `properties.bla is null` and required rewriting `is set` and `is not set` logic.
+
+2. **Including a nullable checkbox in data management.** Similar to the previous option, but required users to input if they expect a property to be null. This required a lot of user work and remains confusing if they didn’t understand how to handle null.
+
+3. **Create JS-like `'bla' (NOT) IN properties` expression logic.** We felt this was too complex for users and makes the question of `is null` for an unset property unclear. 
+
+4. **Use the ClickHouse function `sumOrDefaultIf`.** Faces the same problem of being null if any values are null.
+
+5. **Use non-ANSI compliant string operator.** ANSI is SQL standards for function and syntax. For example, `f'{NULL}-haha-{properties.bla}'` is non-ANSI compliant because it doesn’t conform to standard SQL string concatenation syntax. We decided it is better to keep the standard operators and be flexible with types instead.
+
+## The solution(s) to our null problems
+
+The solution was to handle null for values we expect to be strings differently than numbers.
```

Comment:
```suggestion
The solution was to handle null for values we expect to be strings differently than those that are expected to be numbers.
```

Reply:
This solution is very broad. Our solution was very simple: modify how `concat` works (it's also used if you add strings together with the `||` operator).

Previously pre-hogql everything was an empty string if null, but this is not correct. Ever since we introduced nulls, the biggest painpoint has been with concat-ing values. 

This `concat` function accepts only strings. However users want to join together strings, numbers, nulls and other values. This was the biggest pain point with nulls. So we made concat also accept numbers, nulls, dates and anything else. We didn't change any other function.

Reply:
reworded to be more specific about concat.